### PR TITLE
[Impeller] fail CI if validations are enabled but not available.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/145041

Will test this by unsetting the VVL configuration

### CODE REVIEW

File: impeller/renderer/backend/vulkan/context_vk_unittests.cc
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: Is this compiled into `impeller_golden_tests`?  We should have one of these for that target too.

File: impeller/playground/backend/vulkan/playground_impl_vk.cc
Comment (jonahwilliams) [reply impeller/playground/backend/vulkan/playground_impl_vk.cc:91]: IIRC both goldens and regular playground tests go through this path to create the context.

File: impeller/renderer/backend/vulkan/context_vk_unittests.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: This is just a check that the setting works. I've enabled this setting for playgrounds, so I should be able to confirm that removing any of the VVL config causes failures.

Comment (gaaclarke) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: What I meant is that we want to make sure that `impeller_golden_tests --gtest_filter=ContextVKTest.FatalMissingValidations` is executed too.  I suspect this will execute for `impeller_tests` but not `impeller_golden_tests`.

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: That doesn't matter though, that isn't what is running the check.

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: its just a test that the check works

File: impeller/playground/backend/vulkan/playground_impl_vk.cc
Comment (gaaclarke) [reply impeller/playground/backend/vulkan/playground_impl_vk.cc:91]: I'm not sure.  I'd set a breakpoint in the impeller_golden_tests here to make sure.  We swap out playground tests for golden tests in that target so this might not be getting hit.  That's why we should make sure we have the test for that executable too (as mentioned in the other comment).

Comment (jonahwilliams) [reply impeller/playground/backend/vulkan/playground_impl_vk.cc:91]: Both impeller unittests and goldens have failed on the commit where I broke run_tests.py:

https://github.com/flutter/engine/pull/51378/checks?check_run_id=22626102272

I think that means its working.

File: impeller/renderer/backend/vulkan/context_vk_unittests.cc
Comment (gaaclarke) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: > its just a test that the check works

I know =)

> that isn't what is running the check.

I know, I want this test running in *both* executables.  In `impeller_unittests` and `impeller_golden_tests` we should fail if we don't have vulkan validation.

File: impeller/playground/backend/vulkan/playground_impl_vk.cc
Comment (jonahwilliams) [reply impeller/playground/backend/vulkan/playground_impl_vk.cc:91]: No its in there: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8753556488722177297/+/u/test:_Impeller-golden__dart_and_engine_tests_for_host_release/stdout

It did hit a flake during one of the reruns though.

Comment (gaaclarke) [reply impeller/playground/backend/vulkan/playground_impl_vk.cc:91]: That failure is when executing `impeller_unittests`.  Let's GVC when you have a chance.  I think I can clear this up in under 5 minutes over gvc =)

**Review by matanlurey — COMMENTED** [review]
In general the change looks good to me, but defer to Aaron on the comments he's left already.

**Comment by gaaclarke** [comment]
> In general the change looks good to me, but defer to Aaron on the comments he's left already.

We had a vulkan mind meld yesterday and we're of one minds here.  My suggestion is to get this working for the `impeller_golden_tests` executable too.  If it is involved, filing an issue and addressing in a separate PR is fine.

File: impeller/renderer/backend/vulkan/context_vk_unittests.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:225]: I looked into this more. It doesn't seem possible for me to set up a playground where validations are requested but still misisng, without mocking out the loader. Did you have a different idea here?

**Comment by jonahwilliams** [comment]
okay ... I can't figure out how to do this in the testbed itself.

File: impeller/renderer/backend/vulkan/context_vk_unittests.cc
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/context_vk_unittests.cc:226]: Are you getting a warning about running this in a multithreaded environment?  This works by forking and catching signals which won't work reliably on multi threaded processes.

**Review by gaaclarke — COMMENTED** [review]
lgtm modulo filing an issue for golden test runner and verifying that we aren't getting undefined behavior in EXPECTED_DEATH.

File: impeller/renderer/backend/vulkan/context_vk_unittests.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:226]: Not running locally

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:226]: Is there a cmd line flag I can test this locally with?

Comment (gaaclarke) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:226]: The tests that you execute before your test are going to matter.  If ever there is any test that can spawn a thread and not join it, like creating a thread pool, that run before this test it will be a problem.  I'd just try running the test runner with every test enabled.  Maybe try enabling the test shuffle a few times and seeing if it prints out the error.  I almost guarantee there is at least one test that is leaking threads though.  That's what I was seeing with the golden test runner at least.

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:226]: gtest parallel runs multiple gtest processes, if one or more of those process spawn a subprocess that shouldn't cause any issues. The problems we've had usually stem from reading/writing to the file system.

Comment (gaaclarke) [reply impeller/renderer/backend/vulkan/context_vk_unittests.cc:226]: The problem isn't from running the tests in multiple processes, it comes from the `signal` function.  Which thread will handle a signal is undefined.  That's what's being used to catch the death of the process.  I'm pretty sure that's why googletest prints out the warning.  If you don't see the warning and want to give it a shot that sounds good to me.  We should add a comment to the test in case it ever starts flaking.

**Comment by chinmaygarde** [comment]
I can't tell if we are still making progress on this.

**Comment by jonahwilliams** [comment]
I am waiting on follow up from review from either @chinmaygarde , @gaaclarke  or @matanlurey . I cannot make the requested additional test (i don't know how), so either we can land this with tweaks or I can close the PR.

**Review by chinmaygarde — APPROVED** [review]
So, per my understanding, the ask is to request mocking out the loader to simulate the fault?

I think the patch in this state is an improvement over the current state of things and I'm fine landing this to make forward progress.

**Comment by jonahwilliams** [comment]
Right, I need to mock out the failure and mock out parts of the playground harness. I don't think it would be a great test.

**Comment by matanlurey** [comment]
Apologies for missing this.

Wanna sync for a few minutes tomorrow morning? Either I can convince you there is a reasonable way to test this or concede let's land it as-is.

**Comment by matanlurey** [comment]
You're probably right btw, it's more me wanting to understand the playground harness more.

**Comment by chinmaygarde** [comment]
Was there a sync on this? This feels like a generally valuable thing to have.

**Comment by jonahwilliams** [comment]
There was not. I maintain that 1) the test I am adding is useful and 2) I do not know how to add the second test that Is being requested. I have no plans to work on this further.

**Comment by matanlurey** [comment]
I think this is useful as-is. Let's land it.

**Review by matanlurey — APPROVED** [review]

**Review by gaaclarke — APPROVED** [review]
I validated that the warning isn't being printed out for the 2 instances it's run here: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8751539136032949377/+/u/test:_Host_Tests_for_host_debug/stdout

I would still prefer a comment in case the sheriff runs into flakes otherwise lgtm.

**Pull request merged**
